src/crimson/os/seastore/randomblock_manager.cc

```diff
@@ -0,0 +1,508 @@
+   * 3. link free blocks to onode
+   * Due to in-memory block allocator is the next work to do,
+   * just read the block bitmap directly to find free blocks here.
+   *
+   */
+  return find_free_block(t, size
+      ).safe_then([this, &t, size] (auto allocated) mutable -> allocate_ertr::future<> {
+	logger().debug("after find_free_block: allocated {}", allocated);
+	if (allocated * super.block_size < size) {
+	  t.clear_rbm_allocated_blocks();
+	  return crimson::ct_error::enospc::make();
+	}
+	return allocate_ertr::now();
+	}
+      ).handle_error(
+	allocate_ertr::pass_further{},
+	crimson::ct_error::assert_all{
+	  "Invalid error find_free_block in RandomBlockManager::alloc_extent"
+	}
+	);
+}
+
+RandomBlockManager::free_block_ertr::future<> 
+RandomBlockManager::free_extent(Transaction &t, blk_paddr_t blk_paddr)
+{
+
+  blk_id_t block_no = blk_paddr / super.block_size;
+  auto addr = (super.start_alloc_area +
+	      block_no / max_block_by_bitmap_block()) 
+	      * super.block_size;
+  auto bp = bufferptr(ceph::buffer::create_page_aligned(super.block_size));
+  return device->read(
+      addr,
+      bp
+      ).safe_then([&, this]() {
+	logger().debug("free_extent: addr {}", addr);
+	rbm_bitmap_block_t b_block(super.block_size);
+	bufferlist bl_bitmap_block;
+	bl_bitmap_block.append(bp);
+	encode(b_block, bl_bitmap_block);
+
+	auto remain_offset = block_no % max_block_by_bitmap_block();
+	b_block.clear_bit(remain_offset);
+	bufferlist bl;
+	encode(b_block, bl);
+	logger().debug("free_extent: free block no {}",
+		       convert_bitmap_block_no_to_block_id(remain_offset, addr));
+	std::vector<blk_id_t> ids;
+	ids.push_back(convert_bitmap_block_no_to_block_id(remain_offset, addr));
+	rbm_extent_t extent{
+	  extent_types_t::RBM_ALLOC_INFO,
+	  ids,
+	  addr,
+	  bl
+	};
+	t.add_rbm_allocated_blocks(extent);
+	return free_block_ertr::now();
+	}
+  ).handle_error(
+    free_block_ertr::pass_further{},
+    crimson::ct_error::assert_all{
+      "Invalid error in RandomBlockManager::free_extent"
+    }
+    );
+}
+
+RandomBlockManager::write_ertr::future<>
+RandomBlockManager::complete_allocation(Transaction &t)
+{
+  const auto alloc_blocks = t.get_rbm_allocated_blocks();
+  return seastar::do_with(alloc_blocks, 
+	  [&, this] (auto &alloc_blocks) mutable {
+    return crimson::do_for_each(
+	alloc_blocks,
+	[this, &alloc_blocks](auto &alloc) {
+	  logger().debug(
+	      "complete_allocation: addr {}",
+	      alloc.addr);
+	  bufferptr bptr;
+	  try {
+	    bptr = bufferptr(ceph::buffer::create_page_aligned(alloc.bl.length()));
+	    auto iter = alloc.bl.cbegin();
+	    iter.copy(alloc.bl.length(), bptr.c_str());
+	  } catch (const std::exception &e) {
+	    logger().error(
+		"RandomBlockManager::complete_allocation: "
+		"exception creating aligned buffer {}",
+		e
+		);
+	    ceph_assert(0 == "unhandled exception");
+	  }
+	  return device->write(
+	      alloc.addr,
+	      bptr);
```

Comment:
Can this be failed in the middle due to unexpected device/power failure? If so, could you elaborate how to detect/recover a corrupted bitmap block during restart?

Reply:
Right. In the case you mentioned, CircularJournalManager can cover such case. If the failure occurs during this write, the CircularJournalManager can replay all contents before the failure.

Reply:
OK, does this mean any updated bitmap blocks need to be written into the journal as-a-whole first, and then write into the original location `alloc.addr` during journal trimming, meaning that 1 bit flip might result in 2 x 4K writes?

Reply:
To be clear, the transaction has write-ahead information (alloc_extents iirc) that the journal implementation (maybe CircularJournalManager, maybe not) will encode along with the rest of the transaction.  Upon replay, the journalled updates will be replayed to RandomBlockManager.

Reply:
@cyx1231st That part doesn't seem to exist yet, but there's no reason they couldn't use a delta to encode a more specific mutation.  They are currently using an extent_info_t, but I expect them to change that to a delta.